pad.go

```diff
@@ -0,0 +1,105 @@
+package merkletree
+
+import (
+	"errors"
+
+	"github.com/coniks-sys/libmerkleprefixtree-go/crypto"
+)
+
+var (
+	ErrorSTRNotFound = errors.New("[merkletree] STR not found")
+	ErrorNilPolicies = errors.New("[merkletree] Nil policies")
+)
+
+// PAD is an acronym for persistent authenticated dictionary
+type PAD struct {
+	key          crypto.SigningKey
+	tree         *MerkleTree
+	snapshots    map[uint64]*SignedTreeRoot
+	loadedEpochs []uint64 // slice of epochs in snapshots
+	currentSTR   *SignedTreeRoot
+}
+
+// NewPAD creates new PAD consisting of an array of hash chain
+// indexed by the epoch and its maximum length is len
+func NewPAD(policies Policies, key crypto.SigningKey, len int64) (*PAD, error) {
+	if policies == nil {
+		panic(ErrorNilPolicies)
+	}
+	var err error
+	pad := new(PAD)
+	pad.key = key
+	pad.tree, err = NewMerkleTree()
+	if err != nil {
+		return nil, err
+	}
+	pad.snapshots = make(map[uint64]*SignedTreeRoot, len)
+	pad.loadedEpochs = make([]uint64, 0, len)
+	pad.updateInternal(policies, 1)
+	return pad, nil
+}
+
+// if policies is nil, the previous policies will be used
+func (pad *PAD) generateNextSTR(policies Policies, m *MerkleTree, epoch uint64) {
+	var prevStrHash []byte
+	if pad.currentSTR == nil {
+		prevStrHash = make([]byte, crypto.HashSizeByte)
+	} else {
+		prevStrHash = crypto.Digest(pad.currentSTR.serialize())
```

Comment:
We had imagined ``prevStrHash`` would be ``hash(currentSTR.serialize())`` because it's a hash chain of STRs, not a hash chain of their signatures (per [wikipedia's definition](https://en.wikipedia.org/wiki/Hash_chain)). I suppose this may just be an implementation detail, though?